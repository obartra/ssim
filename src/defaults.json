{
	"windowSize": 11,
	"k1": 0.01,
	"k2": 0.03,
	"bitDepth": 8,
	"downsample": "original",
	"ssim": "fast",
	"maxSize": 256
}
