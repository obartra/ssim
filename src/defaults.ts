import { Options } from './types'

export const defaults: Options = {
  windowSize: 11,
  k1: 0.01,
  k2: 0.03,
  bitDepth: 8,
  downsample: 'original',
  ssim: 'weber',
  maxSize: 256,
  rgb2grayVersion: 'integer',
}
