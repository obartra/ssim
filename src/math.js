/**
 * Computes the mean value of a given array. It is the sum of a list of numbers divided by the
 * number of numbers in the list.
 *
 * @method average
 * @param {Number[]} xn - The target array
 * @returns {Number} average - The mean value of all elements within the array
 * @public
 * @memberOf math
 * @since 0.0.1
 */
function average(xn) {
	return sum(xn) / xn.length;
}

/**
 * Computes the sum of a given array. It adds all values within the array and returns the total
 *
 * @method sum
 * @param {Number[]} xn - The target array
 * @returns {Number} sum - The total value
 * @private
 * @memberOf math
 * @since 0.0.1
 */
function sum(xn) {
	let out = 0;

	for (let x = 0; x < xn.length; x++) {
		out += xn[x];
	}
	return out;
}

/**
 * Computes the largest integer less than or equal to a given number for each member of a given
 * array.
 *
 * @method floor
 * @param {Number[]} xn - The target array
 * @returns {Number[]} floorArr - An array with the Math.floor value for each element of the target
 * array
 * @private
 * @memberOf math
 * @since 0.0.1
 */
function floor(xn) {
	const out = [];

	for (let x = 0; x < xn.length; x++) {
		out[x] = Math.floor(xn[x]);
	}

	return out;
}

/**
 * Computes the sum of all elements within a matrix
 *
 * @method sum2d
 * @param {Array.<Array.<Number>>} mx - The input matrix
 * @returns {Number} sum - The total value of adding each cell
 * @public
 * @memberOf math
 * @since 0.0.2
 */
function sum2d(mx) {
	let out = 0;

	for (let x = 0; x < mx.length; x++) {
		for (let y = 0; y < mx[0].length; y++) {
			out += mx[x][y];
		}
	}

	return out;
}

/**
 * Adds values of two matrices of the same size
 *
 * @method add2dMx
 * @param {Array.<Array.<Number>>} mx1 - The first input matrix
 * @param {Array.<Array.<Number>>} mx2 - The second input matrix
 * @returns {Array.<Array.<Number>>} sumMx - A matrix with a cell-by-cell sum of `mx1` and `mx2`
 * @private
 * @memberOf math
 * @since 0.0.2
 */
function add2dMx(mx1, mx2) {
	const out = [];

	for (let x = 0; x < mx1.length; x++) {
		out[x] = [];
		for (let y = 0; y < mx1[0].length; y++) {
			out[x][y] = mx1[x][y] + mx2[x][y];
		}
	}

	return out;
}

/**
 * Adds a constant value two each matrix cell
 *
 * @method add2dScalar
 * @param {Array.<Array.<Number>>} mx1 - The first input matrix
 * @param {Number} increase - The value to add
 * @returns {Array.<Array.<Number>>} sumMx - The cell-by-cell matrix sum of `mx1` and `increase`
 * @private
 * @memberOf math
 * @since 0.0.2
 */
function add2dScalar(mx, increase) {
	const out = [];

	for (let x = 0; x < mx.length; x++) {
		out[x] = [];
		for (let y = 0; y < mx[0].length; y++) {
			out[x][y] = mx[x][y] + increase;
		}
	}

	return out;
}

/**
 * Adds values of two matrices of the same size or a matrix and a constant
 *
 * @method add2d
 * @param {Array.<Array.<Number>>} mx1 - The first input matrix
 * @param {Array.<Array.<Number>>|Number} increase - The second input matrix or the constant value
 * @returns {Array.<Array.<Number>>} sumMx - A matrix with a cell-by-cell sum of the first and
 * second parameters
 * @public
 * @memberOf math
 * @since 0.0.2
 */
function add2d(mx, increase) {
	if (typeof increase === 'number') {
		return add2dScalar(mx, increase);
	}
	return add2dMx(mx, increase);
}

/**
 * Divides each matrix cell by a constant value
 *
 * @method divide2dScalar
 * @param {Array.<Array.<Number>>} mx1 - The first input matrix
 * @param {Number} divisor - The value to divide by
 * @returns {Array.<Array.<Number>>} dividedMx - The cell-by-cell matrix divison of `mx1` and
 * `divisor`
 * @private
 * @memberOf math
 * @since 0.0.2
 */
function divide2dScalar(mx, divisor) {
	const out = [];

	for (let x = 0; x < mx.length; x++) {
		out[x] = [];
		for (let y = 0; y < mx[0].length; y++) {
			out[x][y] = mx[x][y] / divisor;
		}
	}

	return out;
}

/**
 * Divides, cell-by-cell, values of two matrices of the same size
 *
 * @method divide2dMx
 * @param {Array.<Array.<Number>>} mx1 - The first input matrix
 * @param {Array.<Array.<Number>>} mx2 - The second input matrix
 * @returns {Array.<Array.<Number>>} divideMx - A matrix with a cell-by-cell division of `mx1`/`mx2`
 * @private
 * @memberOf math
 * @since 0.0.2
 */
function divide2dMx(mx1, mx2) {
	const out = [];

	for (let x = 0; x < mx1.length; x++) {
		out[x] = [];
		for (let y = 0; y < mx1[0].length; y++) {
			out[x][y] = mx1[x][y] / mx2[x][y];
		}
	}

	return out;
}

/**
 * Divides values of two matrices of the same size or between a matrix and a constant
 *
 * @method divide2d
 * @param {Array.<Array.<Number>>} mx1 - The first input matrix
 * @param {Array.<Array.<Number>>|Number} divisor - The second input matrix or the constant value
 * @returns {Array.<Array.<Number>>} divideMx - A matrix with the cell-by-cell division of the first
 * and second parameters
 * @public
 * @memberOf math
 * @since 0.0.2
 */
function divide2d(mx, divisor) {
	if (typeof divisor === 'number') {
		return divide2dScalar(mx, divisor);
	}
	return divide2dMx(mx, divisor);
}

/**
 * Multiplies each matrix cell by a constant value
 *
 * @method multiply2dScalar
 * @param {Array.<Array.<Number>>} mx1 - The first input matrix
 * @param {Number} multiplier - The value to multiply each cell with
 * @returns {Array.<Array.<Number>>} multMx - The cell-by-cell matrix multiplication of `mx1`
 * and `multiplier`
 * @private
 * @memberOf math
 * @since 0.0.2
 */
function multiply2dScalar(mx, multiplier) {
	const out = [];

	for (let x = 0; x < mx.length; x++) {
		out[x] = [];
		for (let y = 0; y < mx[0].length; y++) {
			out[x][y] = mx[x][y] * multiplier;
		}
	}

	return out;
}

/**
 * Multiplies, cell-by-cell, values of two matrices of the same size
 *
 * @method multiply2dMx
 * @param {Array.<Array.<Number>>} mx1 - The first input matrix
 * @param {Array.<Array.<Number>>} mx2 - The second input matrix
 * @returns {Array.<Array.<Number>>} multMx - A matrix with a cell-by-cell multiplication of
 * `mx1` * `mx2`
 * @private
 * @memberOf math
 * @since 0.0.2
 */
function multiply2dMx(mx1, mx2) {
	const out = [];

	for (let x = 0; x < mx1.length; x++) {
		out[x] = [];
		for (let y = 0; y < mx1[0].length; y++) {
			out[x][y] = mx1[x][y] * mx2[x][y];
		}
	}

	return out;
}

/**
 * Multiplies values of two matrices of the same size or between a matrix and a constant
 *
 * @method multiply2d
 * @param {Array.<Array.<Number>>} mx1 - The first input matrix
 * @param {Array.<Array.<Number>>|Number} multiplier - The second input matrix or the constant value
 * @returns {Array.<Array.<Number>>} multMx - A matrix with the cell-by-cell multiplication of the
 * first and second parameters
 * @public
 * @memberOf math
 * @since 0.0.2
 */
function multiply2d(mx, multiplier) {
	if (typeof multiplier === 'number') {
		return multiply2dScalar(mx, multiplier);
	}
	return multiply2dMx(mx, multiplier);
}

/**
 * Generates the cell-by-cell square value of a target matrix
 *
 * @method square2d
 * @param {Array.<Array.<Number>>} mx - The target matrix
 * @returns {Array.<Array.<Number>>} squareMx - A matrix with squared value of each cell
 * @public
 * @memberOf math
 * @since 0.0.2
 */
function square2d(mx) {
	return multiply2d(mx, mx);
}

/**
 * Calculates the total mean value for a given matrix
 *
 * @method mean2d
 * @param {Array.<Array.<Number>>} mx - The target matrix
 * @returns {Number} mean - The total mean of each cell
 * @public
 * @memberOf math
 * @since 0.0.2
 */
function mean2d(mx) {
	return sum2d(mx) / (mx.length * mx[0].length);
}

/**
 * Generates all basic arithmetic and matrix computations required
 *
 * Most of these methods use plain for loops and reduce nested calls. This results in about ~100x
 * improvement on SSIM computation for 512x512 images on recent versions of node (~v6.7) over
 * implementations using map or reduce.
 *
 * @namespace math
 */
module.exports = {
	average,
	sum2d,
	add2d,
	divide2d,
	multiply2d,
	square2d,
	mean2d,
	floor
};
