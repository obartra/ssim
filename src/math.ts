/**
 * Generates all basic arithmetic and matrix computations required
 *
 * Most of these methods use plain for loops and reduce nested calls. This results in about ~100x
 * improvement on SSIM computation for 512x512 images on recent versions of node (~v6.7) over
 * implementations using map or reduce.
 *
 * @namespace math
 */
import { Matrix } from './types'

/**
 * Computes the mean value of a given array. It is the sum of a list of numbers divided by the
 * number of numbers in the list.
 *
 * @method average
 * @param {Number[]} xn - The target array
 * @returns {Number} average - The mean value of all elements within the array
 * @public
 * @memberOf math
 * @since 0.0.1
 */
export function average(xn: number[]): number {
  return sum(xn) / xn.length
}

/**
 * Computes the sum of a given array. It adds all values within the array and returns the total
 *
 * @method sum
 * @param {Number[]} xn - The target array
 * @returns {Number} sum - The total value
 * @private
 * @memberOf math
 * @since 0.0.1
 */
export function sum(xn: number[]): number {
  let out = 0

  for (let x = 0; x < xn.length; x++) {
    out += xn[x]
  }
  return out
}

/**
 * Computes the largest integer less than or equal to a given number for each member of a given
 * array.
 *
 * @method floor
 * @param {Number[]} xn - The target array
 * @returns {Number[]} floorArr - An array with the Math.floor value for each element of the target
 * array
 * @private
 * @memberOf math
 * @since 0.0.1
 */
export function floor(xn: number[]): number[] {
  const out = new Array(xn.length)

  for (let x = 0; x < xn.length; x++) {
    out[x] = Math.floor(xn[x])
  }

  return out
}

/**
 * Computes the sum of all elements within a matrix
 *
 * @method sum2d
 * @param {Matrix} A - The input matrix
 * @returns {Number} sum - The total value of adding each cell
 * @public
 * @memberOf math
 * @since 0.0.2
 */
export function sum2d({ data }: Matrix): number {
  let out = 0

  for (let x = 0; x < data.length; x++) {
    out += data[x]
  }

  return out
}

/**
 * Adds values of two matrices of the same size
 *
 * @method add2dMx
 * @param {Matrix} A - The first input matrix
 * @param {Matrix} B - The second input matrix
 * @returns {Matrix} out - A matrix with a cell-by-cell sum of `A` and `B`
 * @private
 * @memberOf math
 * @since 0.0.2
 */
function add2dMx(
  { data: ref1, width, height }: Matrix,
  { data: ref2 }: Matrix
): Matrix {
  const data = new Array(ref1.length)

  for (let x = 0; x < height; x++) {
    const offset = x * width

    for (let y = 0; y < width; y++) {
      data[offset + y] = ref1[offset + y] + ref2[offset + y]
    }
  }

  return {
    data,
    width,
    height,
  }
}

/**
 * Subtracts values of second matrix from the first one. It assumes both matrices are of the same
 * size
 *
 * @method subtract2dMx
 * @param {Matrix} A - The first input matrix
 * @param {Matrix} B - The second input matrix
 * @returns {Matrix} out - A matrix with a cell-by-cell subtraction of `A` minus `B`
 * @private
 * @memberOf math
 * @since 0.0.2
 */
function subtract2dMx(
  { data: ref1, width, height }: Matrix,
  { data: ref2 }: Matrix
): Matrix {
  const data = new Array(ref1.length)

  for (let x = 0; x < height; x++) {
    const offset = x * width

    for (let y = 0; y < width; y++) {
      data[offset + y] = ref1[offset + y] - ref2[offset + y]
    }
  }

  return {
    data,
    width,
    height,
  }
}

/**
 * Adds a constant value two each matrix cell
 *
 * @method add2dScalar
 * @param {Matrix} A - The first input matrix
 * @param {Number} increase - The value to add
 * @returns {Matrix} B - The cell-by-cell matrix sum of `A` and `increase`
 * @private
 * @memberOf math
 * @since 0.0.2
 */
function add2dScalar(
  { data: ref, width, height }: Matrix,
  increase: number
): Matrix {
  const data = new Array(ref.length)

  for (let x = 0; x < ref.length; x++) {
    data[x] = ref[x] + increase
  }

  return {
    data,
    width,
    height,
  }
}

/**
 * Adds values of two matrices of the same size or a matrix and a constant
 *
 * @method add2d
 * @param {Matrix} A - The first input matrix
 * @param {Matrix|Number} increase - The second input matrix or the constant value
 * @returns {Matrix} B - A matrix with a cell-by-cell sum of the first and second parameters
 * @public
 * @memberOf math
 * @since 0.0.2
 */
export function add2d(A: Matrix, increase: Matrix | number): Matrix {
  if (typeof increase === 'number') {
    return add2dScalar(A, increase)
  }
  return add2dMx(A, increase)
}

/**
 * Subtracts values of two matrices of the same size or a matrix and a constant
 *
 * @method subtract2d
 * @param {Matrix} A - The first input matrix
 * @param {Matrix|Number} decrease - The second input matrix or the constant value
 * @returns {Matrix} B - A matrix with a cell-by-cell subtraction of the first parameter minus the
 * second one
 * @public
 * @memberOf math
 */
export function subtract2d(A: Matrix, decrease: Matrix | number): Matrix {
  if (typeof decrease === 'number') {
    return add2dScalar(A, -decrease)
  }
  return subtract2dMx(A, decrease)
}

/**
 * Divides each matrix cell by a constant value
 *
 * @method divide2dScalar
 * @param {Matrix} A - The first input matrix
 * @param {Number} divisor - The value to divide by
 * @returns {Matrix} B - The cell-by-cell matrix divison of `A` and `divisor`
 * @private
 * @memberOf math
 * @since 0.0.2
 */
function divide2dScalar(
  { data: ref, width, height }: Matrix,
  divisor: number
): Matrix {
  const data = new Array(ref.length)

  for (let x = 0; x < ref.length; x++) {
    data[x] = ref[x] / divisor
  }

  return {
    data,
    width,
    height,
  }
}

/**
 * Divides, cell-by-cell, values of two matrices of the same size
 *
 * @method divide2dMx
 * @param {Matrix} A - The first input matrix
 * @param {Matrix} B - The second input matrix
 * @returns {Matrix} out - A matrix with a cell-by-cell division of `A`/`B`
 * @private
 * @memberOf math
 * @since 0.0.2
 */
function divide2dMx(
  { data: ref1, width, height }: Matrix,
  { data: ref2 }: Matrix
): Matrix {
  const data = new Array(ref1.length)

  for (let x = 0; x < ref1.length; x++) {
    data[x] = ref1[x] / ref2[x]
  }

  return {
    data,
    width,
    height,
  }
}

/**
 * Divides values of two matrices of the same size or between a matrix and a constant
 *
 * @method divide2d
 * @param {Matrix} A - The first input matrix
 * @param {Matrix|Number} divisor - The second input matrix or the constant value
 * @returns {Matrix} B - A matrix with the cell-by-cell division of the first and second parameters
 * @public
 * @memberOf math
 * @since 0.0.2
 */
export function divide2d(A: Matrix, divisor: Matrix | number): Matrix {
  if (typeof divisor === 'number') {
    return divide2dScalar(A, divisor)
  }
  return divide2dMx(A, divisor)
}

/**
 * Multiplies each matrix cell by a constant value
 *
 * @method multiply2dScalar
 * @param {Matrix} A - The first input matrix
 * @param {Number} multiplier - The value to multiply each cell with
 * @returns {Matrix} B - The cell-by-cell matrix multiplication of `A` and `multiplier`
 * @private
 * @memberOf math
 * @since 0.0.2
 */
function multiply2dScalar(
  { data: ref, width, height }: Matrix,
  multiplier: number
): Matrix {
  const data = new Array(ref.length)

  for (let x = 0; x < ref.length; x++) {
    data[x] = ref[x] * multiplier
  }

  return {
    data,
    width,
    height,
  }
}

/**
 * Multiplies, cell-by-cell, values of two matrices of the same size
 *
 * @method multiply2dMx
 * @param {Matrix} A - The first input matrix
 * @param {Matrix} B - The second input matrix
 * @returns {Matrix} out - A matrix with a cell-by-cell multiplication of `A` * `B`
 * @private
 * @memberOf math
 * @since 0.0.2
 */
function multiply2dMx(
  { data: ref1, width, height }: Matrix,
  { data: ref2 }: Matrix
): Matrix {
  const data = new Array(ref1.length)

  for (let x = 0; x < ref1.length; x++) {
    data[x] = ref1[x] * ref2[x]
  }

  return {
    data,
    width,
    height,
  }
}

/**
 * Multiplies values of two matrices of the same size or between a matrix and a constant
 *
 * @method multiply2d
 * @param {Matrix} A - The first input matrix
 * @param {Matrix|Number} multiplier - The second input matrix or the constant value
 * @returns {Matrix} out - A matrix with the cell-by-cell multiplication of the first and second
 * parameters
 * @public
 * @memberOf math
 * @since 0.0.2
 */
export function multiply2d(A: Matrix, multiplier: Matrix | number): Matrix {
  if (typeof multiplier === 'number') {
    return multiply2dScalar(A, multiplier)
  }
  return multiply2dMx(A, multiplier)
}

/**
 * Generates the cell-by-cell square value of a target matrix
 *
 * @method square2d
 * @param {Matrix} A - The target matrix
 * @returns {Matrix} B - A matrix with squared value of each cell
 * @public
 * @memberOf math
 * @since 0.0.2
 */
export function square2d(A: Matrix): Matrix {
  return multiply2d(A, A)
}

/**
 * Calculates the total mean value for a given matrix
 *
 * @method mean2d
 * @param {Matrix} A - The target matrix
 * @returns {Number} mean - The total mean of each cell
 * @public
 * @memberOf math
 * @since 0.0.2
 */
export function mean2d(A: Matrix): number {
  return sum2d(A) / A.data.length
}

/**
 * Computes the variance for a given array
 *
 * @method variance
 * @param {Array<Number>} values - The target array
 * @param {Number} [avg=average(values)] - If specified, it will use this values as the average of
 * the array values. If not, it will compute the actual average
 * @returns {Number} varx - The resulting variance value
 * @public
 * @memberOf math
 */
export function variance(
  values: number[],
  avg: number = average(values)
): number {
  let varx = 0
  let i = values.length

  while (i--) {
    varx += (values[i] - avg) ** 2
  }

  return varx / values.length
}

/**
 * Computes the covariance between 2 arrays
 *
 * @method covariance
 * @param {Array<Number>} values1 - The first target array
 * @param {Array<Number>} values2 - The second target array
 * @param {Number} [average1=average(values)] - If specified, it will use this values as the average
 * of the first array. If not, it will compute the actual average
 * @param {Number} [average2=average(values)] - If specified, it will use this values as the average
 * of the second array. If not, it will compute the actual average
 * @returns {Number} cov - The resulting covariance
 * @public
 * @memberOf math
 */
export function covariance(
  values1: number[],
  values2: number[],
  average1: number = average(values1),
  average2: number = average(values2)
): number {
  let cov = 0
  let i = values1.length

  while (i--) {
    cov += (values1[i] - average1) * (values2[i] - average2)
  }

  return cov / values1.length
}
