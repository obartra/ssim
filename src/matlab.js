const { conv2 } = require('./matlab/conv2');
const { filter2 } = require('./matlab/filter2');
const { fspecial } = require('./matlab/fspecial');
const { imfilter } = require('./matlab/imfilter');
const { normpdf } = require('./matlab/normpdf');
const { ones } = require('./matlab/ones');
const { padarray } = require('./matlab/padarray');
const { rgb2gray } = require('./matlab/rgb2gray');
const { skip2d } = require('./matlab/skip2d');
const { sub } = require('./matlab/sub');
const { transpose } = require('./matlab/transpose');
const { zeros } = require('./matlab/zeros');

/**
 * Implements Matlab functions or functionality.
 *
 * The goal here is not a perfect reproduction of matlab logic but just a minimal implementation
 * needed to correctly reproduce the SSIM matlab script.
 *
 * That means that functionality used will be implemented but additional / unused parameters will
 * not.
 *
 * @namespace matlab
 */
module.exports = {
	conv2,
	filter2,
	fspecial,
	imfilter,
	normpdf,
	ones,
	padarray,
	rgb2gray,
	skip2d,
	sub,
	transpose,
	zeros
};
