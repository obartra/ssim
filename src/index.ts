/**
 * SSIM External API
 *
 * @module main
 */
import { rgb2gray, rgb2grayInteger } from './matlab'
import { mean2d } from './math'
import { ssim as fastSsim } from './ssim'
import { originalSsim } from './originalSsim'
import { bezkrovnySsim } from './bezkrovnySsim'
import { downsample } from './downsample'
import { defaults } from './defaults'
import {
  Options,
  Images,
  Matrices,
  Matrix,
  MSSIMMatrix,
  ImageData,
} from './types'
import { weberSsim } from './weberSsim'

export { Options, Matrix }

const ssimTargets = {
  fast: fastSsim,
  original: originalSsim,
  bezkrovny: bezkrovnySsim,
  weber: weberSsim,
}

function validateOptions(options: Options) {
  Object.keys(options).forEach((option) => {
    if (!(option in defaults)) {
      throw new Error(`"${option}" is not a valid option`)
    }
  })
  if ('k1' in options && (typeof options.k1 !== 'number' || options.k1 < 0)) {
    throw new Error(`Invalid k1 value. Default is ${defaults.k1}`)
  }
  if ('k2' in options && (typeof options.k2 !== 'number' || options.k2 < 0)) {
    throw new Error(`Invalid k2 value. Default is ${defaults.k2}`)
  }
  if (!(options.ssim in ssimTargets)) {
    throw new Error(
      `Invalid ssim option (use: ${Object.keys(ssimTargets).join(', ')})`
    )
  }
}

export function getOptions(userOptions?: Partial<Options>): Options {
  const options = { ...defaults, ...userOptions }

  validateOptions(options)

  return options
}

function validateDimensions([pixels1, pixels2, options]: Images): Images {
  if (pixels1.width !== pixels2.width || pixels1.height !== pixels2.height) {
    throw new Error('Image dimensions do not match')
  }

  return [pixels1, pixels2, options]
}

function toGrayScale([pixels1, pixels2, options]: Images): Matrices {
  if (options.rgb2grayVersion === 'original') {
    return [rgb2gray(pixels1), rgb2gray(pixels2), options]
  } else {
    return [rgb2grayInteger(pixels1), rgb2grayInteger(pixels2), options]
  }
}

function toResize([pixels1, pixels2, options]: Matrices): Matrices {
  const pixels = downsample([pixels1, pixels2], options)

  return [pixels[0], pixels[1], options]
}

function comparison([pixels1, pixels2, options]: Matrices): Matrix {
  return ssimTargets[options.ssim](pixels1, pixels2, options)
}

/**
 * @method ssim - The ssim method. You can call the package directly or through the `ssim` property.
 * @public
 * @example import mod = from 'ssim.js';
 * mod(imgBuffer1, imgBuffer2);
 * mod.ssim(imgBuffer1, imgBuffer2);
 */
export function ssim(
  image1: ImageData,
  image2: ImageData,
  userOptions?: Partial<Options>
): {
  ssim_map: Matrix
  mssim: number
  performance: number
} {
  const start = new Date().getTime()
  const options = getOptions(userOptions)
  const ssimMap = comparison(
    toResize(toGrayScale(validateDimensions([image1, image2, options])))
  )
  const mssim =
    (ssimMap as MSSIMMatrix).mssim !== undefined
      ? (ssimMap as MSSIMMatrix).mssim
      : mean2d(ssimMap)
  return {
    mssim,
    ssim_map: ssimMap,
    performance: new Date().getTime() - start,
  }
}

export default ssim
