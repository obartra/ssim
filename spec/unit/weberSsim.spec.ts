import { Options } from "../../src/types";
import { average, covariance, mean2d, variance } from "../../src/math";
import { defaults } from "../../src/defaults";
import { ssim } from "../../src/index";
import { roundTo } from "../helpers/round";
import { sampleCsv, samples } from "../helpers/matrices";
import {
  partialSumMatrix1,
  partialSumMatrix1Forward,
  windowCovariance, windowCovarianceForward, windowMatrix, windowMatrixForward,
  windowSums,
  windowVariance, windowVarianceForward
} from "../../src/weberSsim";
import { sub } from "../../src/matlab";


const testDataImg = {
  width: 4, height: 4, data: [
    0x10, 0x20, 0x30, 0x40,
    0x50, 0x60, 0x70, 0x80,
    0x90, 0xa0, 0xb0, 0xc0,
    0xd0, 0xe0, 0xf0, 0xff
  ]
};
const testDataImg2 = {
  width: 4, height: 4, data: [
    0x0c, 0x1c, 0x2c, 0x3c,
    0x4c, 0x5c, 0x6c, 0x7c,
    0x8c, 0x9c, 0xac, 0xbc,
    0xcc, 0xdc, 0xec, 0xfc
  ]
};

const testPSForwardImg1 = {
  width: 5, height: 5, data: new Int32Array([
    0, 0, 0, 0, 0,
    0, 16, 48, 96, 160,
    0, 96, 224, 384, 576,
    0, 240, 528, 864, 1248,
    0, 448, 960, 1536, 2175])
};


describe("weberSsim", () => {
  let options: Options;
  let k0Options: Options;
  let optionsForward: Options;
  let k0OptionsForward: Options;

  beforeEach(() => {
    options = { ...defaults, ssim: "weber" };
    k0Options = { ...defaults, ssim: "weber", k1: 0, k2: 0 };
    optionsForward = { ...defaults, ssim: "weberForward" };
    k0OptionsForward = { ...defaults, ssim: "weberForward", k1: 0, k2: 0  };
  });

  test("should return same results than Webers's implementation", () => {
    const A = samples["24x18"].gray;
    const B = samples["24x18-degraded"].gray;
    const ssimMap = ssim(A, B, options).ssim_map;
    expect(roundTo(mean2d(ssimMap), 5)).toMatchSnapshot();
  });

  test("rolling ssim mssim should be the same as mean2d ssim", () => {
    const A = sampleCsv.lena;
    const B = sampleCsv.lena02876;
    const { ssim_map: ssimMap, mssim } = ssim(A, B, options);
    expect(roundTo(mean2d(ssimMap), 5)).toMatchSnapshot();
    expect(roundTo(mssim, 5)).toMatchSnapshot();
  });

  test("mean2d to match to equal the Welford's mean of the windowSums", () => {
    const A = sampleCsv.lena;
    const meanAcross = mean2d(A);
    const sums = windowSums(A, 1);
    let welfordMean = 0;
    for (let i = 0; i < sums.data.length; ++i) {
      welfordMean = welfordMean + (sums.data[i] - welfordMean) / (i + 1);
    }
    expect(roundTo(welfordMean, 5)).toBe(roundTo(meanAcross, 5));
  });


  test("matlab averages, variances, covariances should be the same as weber's", () => {
    const A = testDataImg;
    const B = testDataImg2;
    const windowSize = 2;
    const windowWidth = A.width - windowSize + 1;
    const windowHeight = A.height - windowSize + 1;
    const matlabXAverages = [];
    const matlabYAverages = [];
    const matlabXVariances = [];
    const matlabYVariances = [];
    const matlabCovariances = [];
    for (let y = 0; y < windowHeight; ++y) {
      for (let x = 0; x < windowWidth; ++x) {
        const values1 = sub(A, x, windowSize, y, windowSize);
        const values2 = sub(B, x, windowSize, y, windowSize);
        const avgX = average(values1.data);
        const avgY = average(values2.data);
        const varX = variance(values1.data, avgX);
        const varY = variance(values2.data, avgY);
        const covX = covariance(values1.data, values2.data, avgX, avgY);
        matlabXAverages.push(avgX);
        matlabYAverages.push(avgY);
        matlabXVariances.push(varX);
        matlabYVariances.push(varY);
        matlabCovariances.push(covX);
      }
    }
    const sumsX = windowSums(A, windowSize);
    const meansX = Array.from(sumsX.data, (v: number) => v / (windowSize * windowSize));
    const variancesX = windowVariance(A, sumsX, windowSize);
    const sumsY = windowSums(B, windowSize);
    const meansY = Array.from(sumsY.data, (v: number) => v / (windowSize * windowSize));
    const variancesY = windowVariance(B, sumsY, windowSize);
    const covariances = windowCovariance(A, B, sumsX, sumsY, windowSize);
    expect(meansX.length).toBe(matlabXAverages.length);
    expect(meansX.length).toBe(matlabYAverages.length);
    expect(meansX.length).toBe(matlabXVariances.length);
    expect(meansX.length).toBe(matlabYVariances.length);
    expect(meansX.length).toBe(matlabCovariances.length);
    for (let i = 0; i < meansX.length; ++i) {
      expect(meansX[i]).toBe(matlabXAverages[i]);
      expect(meansY[i]).toBe(matlabYAverages[i]);
      expect(variancesX.data[i] / 1024).toBe(matlabXVariances[i]);
      expect(variancesY.data[i] / 1024).toBe(matlabYVariances[i]);
      expect(covariances.data[i] / 1024).toBe(matlabCovariances[i]);
    }
  });

  test("patrialSumMatrixForward should equal the google sheets matrix", () => {
    const sums = partialSumMatrix1Forward(testDataImg, (v => v));
    expect(sums).toEqual(testPSForwardImg1);
  });

  test("make sure windowMatrixBackward and windowMatrixForward produce the same output", () => {
    const sumMatrixBackward1 = partialSumMatrix1(testDataImg, (v => v));
    const sumMatrixForward1 = partialSumMatrix1Forward(testDataImg, (v => v));
    const windowMatrixBackward1 = windowMatrix(sumMatrixBackward1, 2, 1);
    const windowMatrixForward1 = windowMatrixForward(sumMatrixForward1, 2, 1);
    expect(windowMatrixForward1).toEqual(windowMatrixBackward1);

    const sumMatrixBackward2 = partialSumMatrix1(testDataImg2, (v => v));
    const sumMatrixForward2 = partialSumMatrix1Forward(testDataImg2, (v => v));
    const windowMatrixBackward2 = windowMatrix(sumMatrixBackward2, 2, 1);
    const windowMatrixForward2 = windowMatrixForward(sumMatrixForward2, 2, 1);
    expect(windowMatrixForward2).toEqual(windowMatrixBackward2);
  });

  test("make sure windowVarianceForward and windowCovarianceForward match their backward equivalents", () => {
    const windowSize = 2;
    const sumMatrixBackward1 = partialSumMatrix1(testDataImg, (v => v));
    const sumMatrixForward1 = partialSumMatrix1Forward(testDataImg, (v => v));
    const windowMatrixBackward1 = windowMatrix(sumMatrixBackward1, windowSize, 1);
    const windowMatrixForward1 = windowMatrixForward(sumMatrixForward1, windowSize, 1);
    const sumMatrixBackward2 = partialSumMatrix1(testDataImg2, (v => v));
    const sumMatrixForward2 = partialSumMatrix1Forward(testDataImg2, (v => v));
    const windowMatrixBackward2 = windowMatrix(sumMatrixBackward2, windowSize, 1);
    const windowMatrixForward2 = windowMatrixForward(sumMatrixForward2, windowSize, 1);
    const varXForward = windowVarianceForward(testDataImg, windowMatrixForward1, windowSize);
    const varXBackward = windowVariance(testDataImg, windowMatrixBackward1, windowSize);
    const varYForward = windowVarianceForward(testDataImg2, windowMatrixForward2, windowSize);
    const varYBackward = windowVariance(testDataImg2, windowMatrixBackward2, windowSize);
    const covForward = windowCovarianceForward(testDataImg, testDataImg2, windowMatrixForward1, windowMatrixForward2, windowSize);
    const covBackward = windowCovariance(testDataImg, testDataImg2, windowMatrixBackward1, windowMatrixBackward2, windowSize);
    expect(varXForward).toEqual(varXBackward);
    expect(varYForward).toEqual(varYBackward);
    expect(covForward).toEqual(covBackward);
  });

  test("should return NaN with Weber's implementation when k1 / k2 are 0", () => {
    const A = samples["24x18"].gray;
    const B = samples["24x18-degraded"].gray;
    const { ssim_map: ssimMap, mssim } = ssim(A, B, k0Options);
    const { ssim_map: ssimMapForward, mssim: mssimForward } = ssim(A, B, k0OptionsForward);
    expect(mssim).toBeNaN();
    expect(mean2d(ssimMap)).toBeNaN();
    expect(mssimForward).toBeNaN();
    expect(mean2d(ssimMapForward)).toBeNaN();
  });


});
