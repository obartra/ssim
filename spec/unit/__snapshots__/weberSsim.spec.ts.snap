// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`weberSsim rolling ssim mssim should be the same as mean2d ssim 1`] = `0.99327`;

exports[`weberSsim rolling ssim mssim should be the same as mean2d ssim 2`] = `0.99327`;

exports[`weberSsim should return same results than Webers's implementation 1`] = `0.95763`;
